### ISSUE TITLE
adding link can fail for sqlalchemy backend

### ISSUE THREAD

**Issue by lekah** [aiidateam/aiida_core#1088]
The following code in a verdi shell  fails to produce an actual link in the database.
`
n1 = Node().store()
`
`
n2 = Node().store()
`
`
n2._add_dblink_from(n1, link_type=LinkType.INPUT)
`
`
exit # Important: exit here, don't do anything else which might lead to a commit (i.e. store another node)
`
verdi node show [n1.pk] shows no output, also confirmed by checking DB directly.

**Reason**: Node._add_dblink_from calls Node._do_create_link, which adds a link to the session without a commit. If no subsequent commit is issued, the link is not written to the DB.

**Solution**: Add session.commit() in that specific function

**Comment**: On a side-note, this shows the inherent problem of our unit test systems. Within the session, one can query for the existence of this link, since it's in the cache, but it's not yet written to the DB.  which is only remarked if one queries outside the session, in a different process. This is an example why we need  more rigorous testing involving system tests and real  usage examples.   @giovannipizzi  @sphuber  
We might also need to check other dealings with the session which could suffer from a similar problem.

**Comment by szoupanos** [comment]
Thanks for pointing this out! 

This is an easy fix, if this is what we want to do. I.e. commit what it is in the session during the link addition (with, of course, the added link). This is what we do when we call the node.store() in the SQLA backend.

**Comment by sphuber** [comment]
The specific case for `_add_dblink_from` was "fixed" in [this commit](https://github.com/aiidateam/aiida_core/commit/c87a0db7f1f3aa8224dbb782544122216c2e5469). This was necessary, because the tests had to create a link directly on the database level, to avoid the ORM level checks, because it was creating an "illegal" link. The test was then calling a raw SQL statement directly on the database engine to detect this illegal link. Before this fix, this failed of course, exactly for the reason stated by @lekah  as the link was just sitting in the session. This is just to point out that know there is an actual call to `session.commit` in `_add_dblink_from`, but as pointed out by OP, there may be other locations where this behavior is hiding

**Comment by sphuber** [comment]
This is verified to be fixed as of `v1.0.0b1`.
Verified by running:
```
data = Data().store()
Out[4]: <Data: uuid: 23d5aba5-f8c9-484b-afb0-1d4e05f58646 (pk: 1)>
calc = CalculationNode().store()
calc.add_incoming(data, link_label='test', link_type=LinkType.INPUT_CALC)

<EXIT SHELL>

data = load_node(1)
data.get_outgoing().all()
Out[2]: [LinkTriple(node=<CalculationNode: uuid: 85934632-114f-4174-990b-f7c82703216e (pk: 2)>, link_type=<LinkType.INPUT_CALC: 'input_calc'>, link_label=u'test')]
```
